### PR TITLE
Feature/templates extension fields

### PR DESCRIPTION
**Pull request by nk-coding**
- refactors `Template` structure, introduces SubTemplates
  - adds new `SubTemplates` for
    - `InterfacePart`
    - `InterfaceSpecificationVersion`
    - `Interface`
    - `ComponentVersion`
- implementes read part of extension fields of `ExtensibleNode`
- adds `Template` references
- implements read part of templated fields and template schema

### CODE REVIEW

**Review by chriku — APPROVED** [review]

File: src/main/kotlin/gropius/graphql/JSONAwareFunctionDataFetcher.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * @param target if present, the object on which the function is invoked
+ * @param function the function the data fetcher calls
+ * @param applicationContext used to obtain Spring beans
+ */
+class JSONAwareFunctionDataFetcher(
+    target: Any?, function: KFunction<*>, applicationContext: ApplicationContext
+) : SpringDataFetcher(target, function, applicationContext) {
+    override fun mapParameterToValue(param: KParameter, environment: DataFetchingEnvironment): Pair<KParameter, Any?>? {
+        return if (param.findAnnotation<GraphQLType>()?.typeName == "JSON") {
+            param to environment.arguments[param.getGraphQLName() ?: param.name]
```
Comment (ModellbahnFreak) [inline src/main/kotlin/gropius/graphql/JSONAwareFunctionDataFetcher.kt:25]: What is the difference to the super implementation except autowired?

File: src/main/kotlin/gropius/model/template/BaseTemplate.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * @param templateFieldSpecifications the schema for all templated fields defined by this template
+ */
+@DomainNode
+@GraphQLDescription(
+    """Base type for both Template and SubTemplate.
+    Defines templated fields with specific types (defined using JSON schema).
+    """
+)
+abstract class BaseTemplate<T : Node, S : BaseTemplate<T, S>>(
```
Comment (ModellbahnFreak) [inline src/main/kotlin/gropius/model/template/BaseTemplate.kt:27]: Is it possible in kotlin to enforce, that T ALSO needs to be of type TemplatedNode in addition to Node?

**Review by ModellbahnFreak — COMMENTED** [review]
We're going to be so f***** when we have to provide default templates 🤯

File: src/main/kotlin/gropius/graphql/JSONAwareFunctionDataFetcher.kt (hunk 1 above)
Comment (nk-coding) [reply src/main/kotlin/gropius/graphql/JSONAwareFunctionDataFetcher.kt:25]: it does some additional stuff like mapping it to Enums, etc., which we all don't want in this case
see `convertArgumentValue.kt` in graphql-kotlin

File: src/main/kotlin/gropius/model/template/BaseTemplate.kt (hunk 2 above)
Comment (nk-coding) [reply src/main/kotlin/gropius/model/template/BaseTemplate.kt]: possible? yes 
does it provide any benefits in our case? I'm note sure 
does it cause any harm? no 
therefore added it with 24dc068aa37bdd6913a43d6e2c26e558ba8265e5

File: src/main/kotlin/gropius/graphql/JSONAwareFunctionDataFetcher.kt (hunk 1 above)
Comment (ModellbahnFreak) [reply src/main/kotlin/gropius/graphql/JSONAwareFunctionDataFetcher.kt:25]: Pusch. But Whatever

File: src/main/kotlin/gropius/model/template/BaseTemplate.kt (hunk 2 above)
Comment (ModellbahnFreak) [reply src/main/kotlin/gropius/model/template/BaseTemplate.kt]: Nice

**Review by ModellbahnFreak — APPROVED** [review]
Approved

**Pull request merged**